### ISSUE TITLE
Starting Failed with existing database

### ISSUE THREAD

**Issue by aak1247** [bitnami/bitnami-docker-sonarqube#9]
**Description**

Got "Error executing 'postInstallation': ERROR: duplicate key value violates unique constraint "users_login" when starting with existed postgresql database

my docker-compose.yml file:
```yml

version: '2'

services:
### SonarQube Container ###########################
  sonarqube:
    privileged: true
    image: bitnami/sonarqube # sonarqube
    restart: always
    container_name: sonarqube
    hostname: sonarqube
    networks:
      - gitlab_default
    external_links:
      - postgresql
    ports:
      - "9000:9000"
    environment:
      - sonar.jdbc.url=jdbc:postgresql://postgresql:5432/sonar # it's useless but won't harm?
      - sonar.jdbc.username=sonar
      - sonar.jdbc.passowrd=sonar
      - SONARQUBE_JDBC_URL=jdbc:postgresql://postgresql:5432/sonar
      - POSTGRESQL_HOST=postgresql
      - SONARQUBE_DATABASE_NAME=sonar
      - SONARQUBE_DATABASE_USER=sonar
      - SONARQUBE_DATABASE_PASSWORD=sonar
    volumes:
      - /srv/docker/sonar:/bitnami
    #   - /src/docker/sonar/conf:/opt/sonarqube/conf
    #   - /srv/docker/sonar/data:/opt/sonarqube/data
    #   - /srv/docker/sonar/extensions:/opt/sonarqube/extensions
    #   - /srv/docker/sonar/bundled-plugins:/opt/sonarqube/lib/bundled-plugins

networks:
  gitlab_default:
    external: true

```

**Steps to reproduce the issue:**

1. [First Step] change image and variable in docker-compose.yml (I use official image before)
2. [Second Step] docker-compose up
3. [and so on...] Error executing 'postInstallation': ERROR: duplicate key value violates unique constraint "users_login"

**Describe the results you received:**

**Version**

- Output of `docker version`:

```
Docker version 18.05.0-ce, build f150324
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.8.0, build unknown
docker-py version: 1.9.0
CPython version: 2.7.12
OpenSSL version: OpenSSL 1.0.2g  1 Mar 2016
```

**Comment by andresbono** [comment]
Hi @aak1247, can you add the following variable to the environment of the container, initialize it again and share the logs with us?

```
NAMI_LOG_LEVEL=trace
```

It will show more information that will help us to debug the issue.

**Comment by aak1247** [comment]
Tried another image and find a way to use it. So may not try recently. Sorry for that but i'll update the information as long as I have same time to test it.

**Comment by andresbono** [comment]
Thanks for considering it, @aak1247!!

**Comment by kabo** [comment]
I'm getting this error as well.
With standard logging:
```
sonarqu INFO ==> Configuring username...
DETAIL: Key (login)=(admin) already exists.
Error executing 'postInstallation': ERROR: duplicate key value violates unique constraint "users_login"
```

**Comment by kabo** [comment]
Hmm, interesting, running the docker image again I don't get the same error.
Here's the output of `/opt/bitnami/sonarqube/logs/web.log`. It seems to get stuck on `RegisterRules` now?

```
2019.08.21 01:39:07 INFO  web[][o.s.p.ProcessEntryPoint] Starting web
2019.08.21 01:39:08 INFO  web[][o.a.t.u.n.NioSelectorPool] Using a shared selector for servlet write/read
2019.08.21 01:39:09 INFO  web[][o.e.p.PluginsService] no modules loaded
2019.08.21 01:39:09 INFO  web[][o.e.p.PluginsService] loaded plugin [org.elasticsearch.join.ParentJoinPlugin]
2019.08.21 01:39:09 INFO  web[][o.e.p.PluginsService] loaded plugin [org.elasticsearch.percolator.PercolatorPlugin]
2019.08.21 01:39:09 INFO  web[][o.e.p.PluginsService] loaded plugin [org.elasticsearch.transport.Netty4Plugin]
2019.08.21 01:39:11 INFO  web[][o.s.s.e.EsClientProvider] Connected to local Elasticsearch: [127.0.0.1:9001]
2019.08.21 01:39:11 INFO  web[][o.s.s.p.LogServerVersion] SonarQube Server / 7.9.1.27448 / e0c45c97bb2e45d52629f3422ac91718020f6db9
2019.08.21 01:39:11 INFO  web[][o.sonar.db.Database] Create JDBC data source for jdbc:postgresql://xxx.eu-west-1.rds.amazonaws.com/sq
2019.08.21 01:39:20 INFO  web[][o.s.s.p.ServerFileSystemImpl] SonarQube home: /opt/bitnami/sonarqube
2019.08.21 01:39:20 INFO  web[][o.s.s.u.SystemPasscodeImpl] System authentication by passcode is disabled
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin Git / 1.8.0.1574 / aec3dc8f5228aabd218e1cd31ac6e6515a43715d
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin GitHub Authentication for SonarQube / 1.5.0.870 / 153f7c7af7a264adb0fcbe5fee87bdd140a6a3a1
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin JaCoCo / 1.0.2.475 / b79a4724f3a9af1051266b4f8ca0460977295ead
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin LDAP / 2.2.0.608 / 79dc3fa4393a29667673c70182f3016288b548b7
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SAML 2.0 Authentication for SonarQube / 1.1.0.181 / 16fcb5be2d3bcfee8920b29cd758d8b957a18acb
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarC# / 7.15.0.8572 / e0ad49e38a28a8fc333ba746fc998e48678f6a8b
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarCSS / 1.1.1.1010 / 365e21fd0cb9035669fc59f6fec7c8fd28a7303b
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarFlex / 2.5.1.1831 / a0c44437f6abb0feec76edd073f91fec64db2a6c
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarGo / 1.1.1.2000 / 40d55921c7a63b67386a053490d17b6723a46cd5
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarHTML / 3.1.0.1615 / 4181edb5eff5605bec82dc0aa15ecd70eaa5857f
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarJS / 5.2.1.7778 / 49f34eaeaad59868d4353d89b1fc5c02bbe51976
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarJava / 5.13.1.18282 / 568f8ed2349f48e250a9329895b9a870100dfbeb
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarKotlin / 1.5.0.315 / 4ff3a145a58f3f84f1b39846a205a129d742e993
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarPHP / 3.2.0.4868 / ec66bd5f8490677eb0ebae82aa17c2a5d9c0e5e7
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarPython / 1.14.1.3143 / eed7b315b6116fe462a19c771013bf3891c92a97
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarRuby / 1.5.0.315 / 4ff3a145a58f3f84f1b39846a205a129d742e993
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarScala / 1.5.0.315 / 4ff3a145a58f3f84f1b39846a205a129d742e993
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarTS / 1.9.0.3766 / 4a4080b78001a78d758d1d0fa0190fb9496b6f57
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarVB / 7.15.0.8572 / e0ad49e38a28a8fc333ba746fc998e48678f6a8b
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin SonarXML / 2.0.1.2020 / c5b84004face582d56f110e24c29bf9c6a679e69
2019.08.21 01:39:28 INFO  web[][o.s.s.p.ServerPluginRepository] Deploy plugin Svn / 1.9.0.1295 / 942e075773975354e32691a60bfd968065703e04
2019.08.21 01:39:29 INFO  web[][o.s.s.p.d.m.c.PostgresCharsetHandler] Verify that database charset supports UTF8
2019.08.21 01:39:32 INFO  web[][o.s.s.p.w.MasterServletFilter] Initializing servlet filter org.sonar.server.ws.WebServiceFilter@7fcf14fa [pattern=UrlPattern{inclusions=[/api/system/migrate_db.*, ...], exclusions=[/api/properties*, ...]}]
2019.08.21 01:39:32 INFO  web[][o.s.s.a.EmbeddedTomcat] HTTP connector enabled on port 9000
2019.08.21 01:39:41 INFO  web[][o.s.s.p.UpdateCenterClient] Update center: https://update.sonarsource.org/update-center.properties (no proxy)
2019.08.21 01:40:03 INFO  web[][o.s.s.e.IndexCreator] Create index [metadatas]
2019.08.21 01:40:04 INFO  web[][o.s.s.e.IndexCreator] Create type metadatas/metadata
2019.08.21 01:40:04 INFO  web[][o.s.s.e.IndexCreator] Create index [components]
2019.08.21 01:40:05 INFO  web[][o.s.s.e.IndexCreator] Create type components/auth
2019.08.21 01:40:05 INFO  web[][o.s.s.e.IndexCreator] Create index [projectmeasures]
2019.08.21 01:40:06 INFO  web[][o.s.s.e.IndexCreator] Create type projectmeasures/auth
2019.08.21 01:40:06 INFO  web[][o.s.s.e.IndexCreator] Create index [rules]
2019.08.21 01:40:06 INFO  web[][o.s.s.e.IndexCreator] Create type rules/rule
2019.08.21 01:40:06 INFO  web[][o.s.s.e.IndexCreator] Create index [issues]
2019.08.21 01:40:07 INFO  web[][o.s.s.e.IndexCreator] Create type issues/auth
2019.08.21 01:40:07 INFO  web[][o.s.s.e.IndexCreator] Create index [users]
2019.08.21 01:40:07 INFO  web[][o.s.s.e.IndexCreator] Create type users/user
2019.08.21 01:40:07 INFO  web[][o.s.s.e.IndexCreator] Create index [views]
2019.08.21 01:40:08 INFO  web[][o.s.s.e.IndexCreator] Create type views/view
2019.08.21 01:40:09 INFO  web[][o.s.s.s.LogServerId] Server ID: 61EE99E8-AWx6IgJdWT40YwBNuZXL
2019.08.21 01:40:11 WARN  web[][o.s.s.a.LogOAuthWarning] For security reasons, OAuth authentication should use HTTPS. You should set the property 'Administration > Configuration > Server base URL' to a HTTPS URL.
2019.08.21 01:40:14 WARN  web[][o.s.a.s.w.WebService$Action] The response example is not set on action api/plugins/download
2019.08.21 01:40:14 WARN  web[][o.s.a.s.w.WebService$Action] The response example is not set on action api/permissions/search_templates
2019.08.21 01:40:18 INFO  web[][o.s.s.t.TelemetryDaemon] Sharing of SonarQube statistics is enabled.
2019.08.21 01:40:18 INFO  web[][o.s.s.n.NotificationDaemon] Notification service started (delay 60 sec.)
2019.08.21 01:40:18 INFO  web[][o.s.s.s.GeneratePluginIndex] Generate scanner plugin index
2019.08.21 01:40:18 INFO  web[][o.s.s.s.RegisterPlugins] Register plugins
2019.08.21 01:40:21 INFO  web[][o.s.s.s.RegisterMetrics] Register metrics
2019.08.21 01:40:59 INFO  web[][o.s.s.r.RegisterRules] Register rules
```

**Comment by kabo** [comment]
Can we re-open this issue please? I don't know how to proceed with troubleshooting. Ping @andresbono

**Comment by andresbono** [comment]
Hi @kabo, did you try with the `NAMI_LOG_LEVEL` environment variable? https://github.com/bitnami/bitnami-docker-sonarqube/issues/9#issuecomment-494295635

**Comment by kabo** [comment]
Yes, I did, but as I wrote
> Hmm, interesting, running the docker image again I don't get the same error.

How does it know to initialize the database or not? I've provided it with credentials to a known working database, and it seems to have tried to initialize it again. Is it looking for some value in the database and if it's not there it initializes the database?

**Comment by andresbono** [comment]
Hi @kabo, the container doesn't check if the database is already initialized or not, that is something that the application should handle.

What we do to differentiate the first run and the rest is storing a file in the persisting volume: `/bitnami/sonarqube/.initialized`.

---

Was the database already populated when you started the container?

This is the docker-compose file that we are using to test the "use existing database" functionality:

Note that the container supports using an existing _but empty_ database.

```yaml
version: '2'

services:
  postgresql:
    image: 'bitnami/postgresql:10'
    environment:
      - POSTGRESQL_USERNAME=sonar
      - POSTGRESQL_DATABASE=sonar
      - POSTGRESQL_PASSWORD=sonar
    volumes:
      - 'postgresql_data:/bitnami/postgresql'
  sonarqube:
    image: bitnami/sonarqube
    ports:
      - "9000:9000"
    environment:
      - POSTGRESQL_HOST=postgresql
      - SONARQUBE_DATABASE_NAME=sonar
      - SONARQUBE_DATABASE_USER=sonar
      - SONARQUBE_DATABASE_PASSWORD=sonar
    volumes:
      - 'sonarqube_data:/bitnami'

volumes:
  sonarqube_data:
    driver: local
  postgresql_data:
    driver: local
```

**Comment by kabo** [comment]
Hmm, ok... We're using this in AWS Fargate, which doesn't permit you to attach any persistent storage. So when we take down a container that was working with the database and set up a brand new empty container we figured it would just connect to the existing database and keep trucking.
Would it be possible to implement this use case? Perhaps by having it check for an environment variable (e.g. `SONARQUBE_DATABASE_INITIALIZED`) as well as checking `/bitnami/sonarqube/.initialized`?

**Comment by andresbono** [comment]
We can evaluate it. In the meantime, can you try the following? Build your own docker image as follows:

```Dockerfile
FROM bitnami/sonarqube:latest

RUN mkdir -p /bitnami/sonarqube && touch /bitnami/sonarqube/.initialized
```

Then, run it on AWS Fargate. Does it work for you? Would it cover your use case?

**Comment by kabo** [comment]
Yup, that does seem to work. Thanks!
Would be neat to not have to build a new image every time you release a new image though. So that environment variable option would be handy :)

**Comment by andresbono** [comment]
Nice, @kabo!

We will leave this issue open until we decide what to do with the `SONARQUBE_DATABASE_INITIALIZED` env-var.

**Comment by goolzerg** [comment]
any news?

**Comment by goolzerg** [comment]
> Nice, @kabo!
> 
> We will leave this issue open until we decide what to do with the `SONARQUBE_DATABASE_INITIALIZED` env-var.

any news?

**Comment by andresbono** [comment]
I'm sorry, @goolzerg, but the internal task is still in our backlog. However, we have made many changes in the image. Are you still facing the same issue?

**Comment by marcosbc** [comment]
Hi @mplescano, we just released [`SonarQube 9.0.0-10-debian-10-r0`](https://github.com/bitnami/bitnami-docker-sonarqube#900-debian-10-r0) which includes major changes, and the source code is now available in the [`rootfs` folder](https://github.com/bitnami/bitnami-docker-sonarqube/tree/master/9/debian-10/rootfs) in this repo.

As part of these changes, we have added a new environment variable `SONARQUBE_SKIP_BOOTSTRAP`, as described in the [Environment variables](https://github.com/bitnami/bitnami-docker-sonarqube#user-and-site-configuration) section in the README, which should allow to initialize with an existing database.

Therefore I'm going to close this issue as resolved.
